### PR TITLE
sql: fix deterministic ID generation

### PR DESCRIPTION
**Pull request by ajwerner**
The code was fooling itself: we often did not re-initialize the eval context
and thus we were using the old non-transactional generator.

Release justification: Only affects testing code.
Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
The numbers go down because there were statements that would fail but would have incremented the ID generator sequence because it was non-transactional.

**Comment by ajwerner** [comment]
@chengxiong-ruan PTAL, I reworked some of the UDF tests to avoid hard-coded IDs as inputs

File: pkg/sql/logictest/testdata/logic_test/udf [hunk 1]
```diff
@@ -1178,28 +1178,56 @@ CREATE FUNCTION f_test_sc(INT) RETURNS INT LANGUAGE SQL AS $$ SELECT 2 $$;
 CREATE SCHEMA test_alter_sc;
 CREATE FUNCTION test_alter_sc.f_test_sc() RETURNS INT LANGUAGE SQL AS $$ SELECT 3 $$;
 
+statement ok
+CREATE FUNCTION get_function_id(namespace STRING, name STRING, argmodes STRING[])
+  RETURNS INT
+  LANGUAGE SQL
+  AS $$
+SELECT oid::INT8 - 100000
+  FROM pg_proc
```
Comment (chengxiong-ruan) [inline pkg/sql/logictest/testdata/logic_test/udf:1187]: This is an excellent idea!!!

**Review by chengxiong-ruan — APPROVED** [review]
💯

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**